### PR TITLE
GH-766 Adds C3 2nd Y-Axis capability

### PR DESCRIPTION
**Pull request by jbrundage**
Fixes GH-766

@GordonSmith @mlzummo @dtsnell4 Please review

http://rawgit.com/jbrundage/Visualization/GH-766/demos/dermatology.html?src/c3chart/Combo.multiY

http://rawgit.com/jbrundage/Visualization/GH-766/demos/dermatology.html?src/c3chart/Column.multiY

http://rawgit.com/jbrundage/Visualization/GH-766/demos/dermatology.html?src/c3chart/Area.multiY

Signed-off-by: jay <jbrundage372@gmail.com>

### CODE REVIEW

**Comment by jbrundage** [comment]
This implementation simply adds **y2** array-type publish parameter. The "columns" that should belong to the Y2 axis are specified in the y2 publish parameter array.

File: src/c3chart/CommonND.js [hunk 1]
```diff
@@ -108,7 +110,7 @@
             },
             y: {
                 show: this.showYGrid(),
-            }
+            },
```
Comment (GordonSmith) [inline src/c3chart/CommonND.js]: remove trailing comma

File: src/c3chart/CommonND.js [hunk 2]
```diff
@@ -108,7 +110,7 @@
             },
             y: {
                 show: this.showYGrid(),
```
Comment (GordonSmith) [inline src/c3chart/CommonND.js]: remove trailing comma - I would have though lint would have caught these?

Comment (jbrundage) [inline src/c3chart/CommonND.js]: I don't think lint checks for trailing commas

**Comment by mlzummo** [comment]
@jbrundage is there a way to configure the positions of the Axes?

**Comment by mlzummo** [comment]
not sure if this is a general issue but looks like the default behavior for c3 selection is multi selection?

**Comment by mlzummo** [comment]
also should have capability of coloring the axis line the same color as the datapoints they belong to or some kind of legend?

**Comment by dtsnell4** [comment]
As for configuring the axes, my assumption was just changing the y2 publish param to "C1" would do it, but that has no affect on the widget at all

**Comment by mlzummo** [comment]
@dtsnell4 im not sure i follow what your saying

File: src/c3chart/CommonND.js [hunk 3]
```diff
@@ -133,6 +135,23 @@
...
+            if(this.columns().length > 1){
+                var cols = this.columns();
+                for(var i = 0;i<y2arr.length;i++){
+                    for(var j = 1;j<cols.length;j++){
+                        axesObj[cols[j]] = y2arr.indexOf(cols[j]) !== -1 ? "y2" : "y";
+                    }
+                }
+                this._config.data.axes = axesObj;
+            }
+        }
```
Comment (dtsnell4) [inline src/c3chart/CommonND.js]: @mlzummo Sorry, I mean changing the y2 publish param in the property editor.   This array is what determines what is associated with the second axis.  In the test data that is [c2, c3] (second and third columns).  I am assuming changing this to [c1] would associate the first column of data with the second y axis.  

The second point of my comment was that doing this in the _Property Editor_ has not effect, which is probably an issue of the Property editor itself, not the chart, but I don't know that for sure.

File: src/c3chart/CommonND.js [hunk 4]
```diff
@@ -133,6 +135,23 @@
             break;
         }
 
+        if(this.y2() instanceof Array && this.y2().length > 0){
```
Comment (mlzummo) [inline src/c3chart/CommonND.js]: if y2 is of type array we dont need to check if its an array?  (though i know sometimes we set things null)

**Comment by mlzummo** [comment]
should we basing the second y axis off the columns or should we use the whole data row?

**Comment by mlzummo** [comment]
same goes for combo chart ...?

File: src/c3chart/CommonND.js (hunk 3 above)
Comment (jbrundage) [inline src/c3chart/CommonND.js]: @dtsnell4 It's an issue with the chart. As far as I'm aware... C3 doesn't allow adding a second Y axis to a chart that's already initialized.

File: src/c3chart/CommonND.js (hunk 4 above)
Comment (jbrundage) [inline src/c3chart/CommonND.js]: Yeah, I just coded that with "null" in mind... to prevent this.y2().length from breaking

**Comment by GordonSmith** [comment]
@mlzummo, @jbrundage can you compare the test factory instantiation code for your two implementations and see how similar the API calls are (I would expect them to be the same)?

File: test/c3chartFactory.js [hunk 5]
```diff
@@ -37,6 +46,15 @@
...
                 });
+            },
+            multiY: function (callback) {
+                require(["test/DataFactory", "src/c3chart/Column"], function (DataFactory, Widget) {
+                    callback(new Widget()
+                        .y2(["C2","C3"])
+                        .columns(DataFactory.ND.multiY.columns)
+                        .data(DataFactory.ND.multiY.data)
+                    );
+                });
```
Comment (jbrundage) [inline test/c3chartFactory.js]: @GordonSmith Here is the API call for c3 column

Comment (jbrundage) [inline test/c3chartFactory.js]: @GordonSmith And here is the API call for AM column
https://github.com/hpcc-systems/Visualization/blob/master/test/amchartFactory.js#L48-L65

**Comment by mlzummo** [comment]
@jbrundage is this the final code?

**Comment by jbrundage** [comment]
@mlzummo I suppose the yAxis publish parameters should be applied to the y2 axis as well... though this differs from the AM Charts implementation

**Comment by mlzummo** [comment]
You can add the axis obj logic to c3 charts

**Pull request closed without merging**